nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/RunMongoAggregation.java

```diff
@@ -0,0 +1,203 @@
+        List<Bson> result = new ArrayList<>();
+
+        ObjectMapper mapper = new ObjectMapper();
+        List<Map> values = mapper.readValue(query, List.class);
+        for (Map val : values) {
+            result.add(new BasicDBObject(val));
+        }
+
+        return result;
+    }
+
+    public static final Validator AGG_VALIDATOR = (subject, value, context) -> {
+        final ValidationResult.Builder builder = new ValidationResult.Builder();
+        builder.subject(subject).input(value);
+
+        if (context.isExpressionLanguageSupported(subject) && context.isExpressionLanguagePresent(value)) {
+            return builder.valid(true).explanation("Contains Expression Language").build();
+        }
+
+        String reason = null;
+        try {
+            buildAggregationQuery(value);
+        } catch (final RuntimeException | IOException e) {
+            reason = e.getLocalizedMessage();
+        }
+
+        return builder.explanation(reason).valid(reason == null).build();
+    };
+
+    static final PropertyDescriptor QUERY = new PropertyDescriptor.Builder()
+            .name("mongo-agg-query")
+            .displayName("Query")
+            .expressionLanguageSupported(true)
+            .description("The aggregation query to be executed.")
+            .required(true)
+            .addValidator(AGG_VALIDATOR)
+            .build();
+
+    static {
+        List<PropertyDescriptor> _propertyDescriptors = new ArrayList<>();
+        _propertyDescriptors.addAll(descriptors);
+        _propertyDescriptors.add(CHARSET);
+        _propertyDescriptors.add(QUERY);
+        _propertyDescriptors.add(QUERY_ATTRIBUTE);
+        _propertyDescriptors.add(BATCH_SIZE);
+        _propertyDescriptors.add(RESULTS_PER_FLOWFILE);
+        _propertyDescriptors.add(SSL_CONTEXT_SERVICE);
+        _propertyDescriptors.add(CLIENT_AUTH);
+        propertyDescriptors = Collections.unmodifiableList(_propertyDescriptors);
+
+        final Set<Relationship> _relationships = new HashSet<>();
+        _relationships.add(REL_RESULTS);
+        _relationships.add(REL_ORIGINAL);
+        _relationships.add(REL_FAILURE);
+        relationships = Collections.unmodifiableSet(_relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return propertyDescriptors;
+    }
+
+    static String buildBatch(List batch) {
+        ObjectMapper mapper = new ObjectMapper();
+        String retVal;
+        try {
+            retVal = mapper.writeValueAsString(batch.size() > 1 ? batch : batch.get(0));
+        } catch (Exception e) {
+            retVal = null;
+        }
+
+        return retVal;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
```

Comment:
It is possible here that there is a flow file in the queue but if there are multiple concurrent tasks, then one of the tasks may not get a flow file. However then some of the concurrent tasks will be behaving in a more event-driven fashion (based on an incoming flow file) and some of the tasks will be behaving in a scheduled fashion. This could get quite confusing in terms of expected behavior and can complicate debugging.

If you are looking to have the processor behave in a more event-driven fashion when flow files are present, then you can check to see if there are any non-loop connections and react accordingly. For example, if there are non-loop connections into this processor, then you can require that session.get() return a flow file, and if there are no incoming connections, you don't have to call session.get() at all.  Check GenerateTableFetch for an example of how the two behaviors are implemented.

Reply:
Ok. I'll take a look at that, and when I've got a solution I'll push an update.

Reply:
I assume you mean this:

```
FlowFile fileToProcess = null;
if (context.hasIncomingConnection()) {
    fileToProcess = session.get();

    if (fileToProcess == null) {
        // Incoming connection with no flow file available, do no work (see capability description)
        return;
    }
}
```

Seems easy enough.